src/blog/speeding-up-v8-heap-snapshots.md

```diff
@@ -0,0 +1,190 @@
+
+```
+ValueBits := 24
+Mask := (1 << ValueBits) - 1  /* 0xffffff */
+OriginalHash(x) := (digits(x) << ValueBits) | (x & Mask)
+```
+
+| x | OriginalHash(x) |
+| -: | -: |
+| 0   | `0x1000000` |
+| 1   | `0x1000001` |
+| 2   | `0x1000002` |
+| 3   | `0x1000003` |
+| 10  | `0x200000a` |
+| 11  | `0x200000b` |
+| 100 | `0x3000064` |
+
+This function was problematic. Some examples of problems with this hash function:
+
+- Once we inserted a string whose hash key value was a small number, we would run into collisions when we tried to store another number in that location, and there would be similar collisions if we tried to store subsequent numbers consecutively.
+- Or even worse: if there were already a lot of consecutive numbers stored in the map, and we wanted to insert a string whose hash key value was in that range, we had to move the entry along all the occupied locations to find a free location.
+
+What did I do to fix it? As the problem comes mostly from numbers represented as strings that would fall in consecutive positions, I modified the hash function so we would rotate the resulting hash value 2 positions to the left.
+
+```
+NewHash(x) := OriginalHash(x) << 2
+``````
+
+| x | OriginalHash(x) | NewHash(x) |
+| -: | -: | -: |
+| 0   | `0x1000000` | `0x4000000` |
+| 1   | `0x1000001` | `0x4000004` |
+| 2   | `0x1000002` | `0x4000008` |
+| 3   | `0x1000003` | `0x400000c` |
+| 10  | `0x200000a` | `0x8000028` |
+| 11  | `0x200000b` | `0x800002c` |
+
+So, for each number, we would introduce 3 free positions. Why rotating 2 bits? Empirical testing across several work-sets showed this number was the best choice to minimize collisions.
+
+[This hashing fix](https://chromium-review.googlesource.com/c/v8/v8/+/4428811) has landed in V8.
+
+## Second Optimization: Caching source positions
+
+After fixing the hashing, we re-profiled and found a further optimization opportunity that would reduce a significant part of the overhead.
+
+For each allocation in the V8 heap, the heap snapshot tries to record the call stack responsible for the allocation. Therefore, for each stack frame, it needs to know the function name and its source location (filename, line number, column number). It was the fetching of this information that turned out to be super slow!
+
+What was happening? It was something similar to [what I fixed in the ETW stack walk](https://chromium-review.googlesource.com/c/v8/v8/+/3669698) and that I explained in [this post](https://blogs.igalia.com/dape/2022/12/21/native-call-stack-profiling-3-3-2022-work-in-v8/).
+
+To compute the source code lines of a function, V8 knows the linear position of the function in the script. But, to find the source line, it needs to traverse the whole script to identify where each newline occurs. This is expensive.
```

Comment:
It's a bit difficult for me to understand what this is explaining, even though I already kind of know how the code works, I think mentioning the line and column numbers here would help:

```suggestion
V8 usually only stores where the the function starts and ends in the form of linear offsets in the script. To calculate line and column numbers for allocation sites within a function, it needs to traverse the whole script and record where the line breaks are.
```

And maybe this can be merged into the `For each allocation in the V8 heap..` paragraph, so that we don't have to repeat "this turns out to be slow" and "this is expensive".

Reply:
I agree. It is quite redundant. I will merge them.